Fixed compilation warnings and more clear Unknown values

mjonss opened the pull request:
I just found your repo very useful for my Remeha Avanta, and just wanted to share back a bit.

mjonss committed:
Fix compilation and cppcheck warnings

mjonss committed:
convert.pl: Add expressions used for AvantaV1_P5

mjonss committed:
Add raw value in brackets in unknown code strings

ribbons committed:
Fix building under Arch Linux

Compilation fails under Arch Linux due to missing includes that don't
cause an issue on other systems, presumably either due to system header
differences or just the updated toolchain.  Add the missing includes to
resolve this.

ribbons commented on convert.pl:
```diff
@@ -29,16 +30,18 @@ sub uint_from_bits
     return 'uint32_t' if ($bits <= 32);
     return 'uint64_t' if ($bits <= 64);
-    die 'Value too large';
+    confess 'Value too large for bits: '. $bits;
```
I'm not quite sure why you've switched from die to confess here as the function is only called in one place?

mjonss replied:
I hit a field that was 80 bit wide, so that it broke and I just wanted to see where it came from. Leaving it with a simple die is ok, I just did not read the code thoroughly enough. (I don't remember which config file I used for hitting the 80 bit wide type :( )

ribbons commented on convert.pl:
```diff
@@ -29,16 +30,18 @@ sub uint_from_bits
 use constant EXPRESSIONS => {
     'A'                     => ['uint8_t',   8, '$id',           1],
     'A x 0.1'               => ['uint8_t',   8, '$id',         0.1],
+    'A x 100'               => ['uint8_t',   8, '$id',         100],
     'A.0 + B.1'             => ['uint16_t', 16, '$id',           1],
     '(A.1 + B.0) x 8'       => ['uint16_t', 16, 'ntohs($id)',    8],
+    '(A.1 + B.0) x 10'      => ['uint16_t', 16, 'ntohs($id)',   10],
```
Out of interest (and so I can reference it in the commit message for these additions), which of the config xml files did you use for your boiler?

mjonss replied:
`perl convert.pl ~/remeha/config/AvantaV1_P5.xml ~/remeha/language/language-en.xml parameter` and `perl convert.pl ~/remeha/config/AvantaV1_P5.xml ~/remeha/language/language-en.xml counter` needed them.
I just noticed that 'counter' would also need "(A.1 + B.0) x 1" and "(A.1 + B.0) x 2". Will update with a PR when I have hacked a bit more.

ribbons reviewed (commented):
I'm so pleased that you found this code useful and thank you for your contributions :smile:.

I have taken the liberty of splitting the changes out into separate commits and adding a commit of my own to fix some more compilation errors.  Only parts of your PR I wanted to check with you are the changes to convert.pl - I've add a couple of questions as inline comments.

ribbons committed:
Make convert.pl handle >64 bits of padding at once

Some of the messages returned by boilers (e.g. parameters) contain more
than 64 bits of continuous unused space.  As there isn't guaranteed to
be a type larger than 64 bits available to use as padding, use as many
uint64_t entries as necessary, followed by the final type based on the
number of bits remaining.

ribbons commented:
Perfect - cheers for your responses, they were very useful.  I've extended convert.pl slightly so it should be able to handle as many bits of padding as are required (I found that the `parameter` message is the one with 80 bits of padding).

Thanks again for the PR and happy hacking :smile:
